internal/tests/api/targets/target_test.go

```diff
@@ -379,6 +380,166 @@ func TestList(t *testing.T) {
 	assert.Equal(filterItem.Id, ul.Items[0].Id)
 }
 
+func TestTarget_AddressMutualExclusiveRelationship(t *testing.T) {
+	tc := controller.NewTestController(t, nil)
+	t.Cleanup(tc.Shutdown)
+
+	client := tc.Client()
+	at := tc.Token()
+	client.SetToken(at.Token)
+	_, proj := iam.TestScopes(t, tc.IamRepo(), iam.WithUserId(at.UserId))
+	tClient := targets.NewClient(client)
+
+	// Create target with a network address association
+	targetResp, err := tClient.Create(tc.Context(), "tcp", proj.GetPublicId(),
+		targets.WithName("test-address"), targets.WithAddress("localhost"), targets.WithTcpTargetDefaultPort(22))
+	require.NoError(t, err)
+	require.NotNil(t, targetResp)
+	require.Equal(t, "localhost", targetResp.GetItem().Address)
+
+	// Setup host catalog, host set, & host resources
+	hc, err := hostcatalogs.NewClient(client).Create(tc.Context(), "static", proj.GetPublicId())
+	require.NoError(t, err)
+	require.NotNil(t, hc)
+	hs, err := hostsets.NewClient(client).Create(tc.Context(), hc.Item.Id)
+	require.NoError(t, err)
+	require.NotNil(t, hs)
+	h, err := hosts.NewClient(client).Create(tc.Context(), hc.Item.Id, hosts.WithStaticHostAddress("localhost"))
+	require.NoError(t, err)
+	require.NotNil(t, h)
+	hUpdate, err := hostsets.NewClient(client).AddHosts(tc.Context(), hs.Item.Id, hs.Item.Version, []string{h.GetItem().Id})
+	require.NoError(t, err)
+	require.NotNil(t, hUpdate)
+
+	// Expect error when associating a host source to the target with a direct network address association
+	targetId := targetResp.GetItem().Id
+	version := targetResp.GetItem().Version
+	updateResp, err := tClient.AddHostSources(tc.Context(), targetId, version, []string{hs.Item.Id})
+	require.Error(t, err)
```

Comment:
Can we verify that this is the expected `409 Conflict` error?

Reply:
The expected status code value is a 400 because grpc-gateway maps FailedPrecondition check errors into bad requests. Here is an example of the output from the cli:
```
Error from controller when performing add-host-sources on target

Error information:
  Kind:                FailedPrecondition
  Message:             target.(Repository).AddTargetHostSources: error creating sets: db.DoTx: target.(Repository).AddTargetHostSources: unable
  to add host sources because a network address is directly assigned to the given target: integrity violation: error #409
  Status:              400
  context:             Error from controller when performing add-host-sources on target
```

Reply:
Ok, fair, but can't we verify that here then? We should be able to parse this error here like we are in the e2e tests, right?

Reply:
Ah, nevermind, I see this has been added now!